### PR TITLE
Add docs for factory generation of artifact names

### PR DESCRIPTION
**Pull request by strickvl**
Relates to https://github.com/zenml-io/zenml/issues/2685 and general docs improvements.

Added the document to the `build-pipelines` section of the 'How To' guides.

### CODE REVIEW

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+---
+description: How to dynamically generate steps and artifacts in your pipelines.
+---
+
+# Dynamically Generating Steps and Artifacts
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:5]: I think maybe the framing of this is "Dynamically assign artifact names" . The dynamic step generation just happens to be a byproduct

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+3. Output artifacts are saved with the same name and incremented version numbers.
+
+For example, when using a preprocessor step that needs to transform train, validation, and test data separately, you might end up with three versions of an artifact called `transformed_data`, making it difficult to track which is which.
+
+ZenML offers two possible ways to address this problem:
+
+1. Using factory functions to create dynamic steps with custom artifact names.
+2. Using metadata to identify artifacts in a single step.
+
+## Using Factory Functions for Dynamic Step Creation
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:27]:
```suggestion
## 1. Using Factory Functions for Dynamic Step Creation
```

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    train_step = create_dynamic_step("train")
+    val_step = create_dynamic_step("validation")
+    test_step = create_dynamic_step("test")
+
+    train_step(train_data)
+    val_step(val_data)
+    test_step(test_data)
+```
+
+This method generates unique artifact names for each step, making it easier to track and retrieve specific artifacts later in your workflow.
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:57]: can you add a screenshot of the DAG produced here?

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from typing_extensions import Annotated
+from zenml import step, get_step_context
+
+@step
+def generic_step(data: Any, prefix: str) -> Annotated[Dict[str, Any], "generic_artifact"]:
+    result = {"processed_data": data}
+
+    # Add custom metadata
+    step_context = get_step_context()
+    step_context.add_output_metadata(
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:74]: I didnt even know ths was a function actually! :D can we see if it really exists? maybe I hallucinated it in my original comment

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+If you prefer using a single step and differentiating artifacts through metadata, try this approach:
+
+```python
+from typing import Any, Dict
+from typing_extensions import Annotated
+from zenml import step, get_step_context
+
+@step
+def generic_step(data: Any, prefix: str) -> Annotated[Dict[str, Any], "generic_artifact"]:
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:69]: maybe it should be called "dataset"
```suggestion
def generic_step(data: Any, prefix: str) -> Annotated[Dict[str, Any], "dataset"]:
```

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    return result
+
+@pipeline
+def metadata_artifact_pipeline(train_data, val_data, test_data):
+    generic_step(train_data, prefix="train")
+    generic_step(val_data, prefix="validation")
+    generic_step(test_data, prefix="test")
+```
+
+This method uses a single `generic_step` but adds custom metadata to each artifact. You can later use this metadata to identify and differentiate between artifacts:
```
Comment (htahir1) [inline docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:88]: might be nice to add a screenshot of the DAG with the metadata highlighted?

**Review by htahir1 — CHANGES_REQUESTED** [review]

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md (hunk 4 above)
Comment (strickvl) [reply docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:99]: yes it does exist :)

File: docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+# Create the dynamic steps
+train_step = create_step(prefix="train")
+validation_step = create_step(prefix="validation")
+test_step = create_step(prefix="test")
+
+# Resolve the steps
+train_step.resolve()
+validation_step.resolve()
+test_step.resolve()
```
[the remark this answers is not in the document]
Comment (strickvl) [reply docs/book/how-to/build-pipelines/dynamically-generate-steps-artifacts.md:54]: Had a discussion with @schustmi about this. It was the only way to make this work. The second of the two options we present is a bit more stable, but if you want to dynamically name the artifacts (vs just adding metadata) then this was what you have to do. The version you posted on the issue didn't work. (Come to think of it, I'll add a comment there updating with the working examples).

**Pull request merged**